### PR TITLE
Rename Theano to Aesara

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR renames Theano to Aesara, changes the Theano-PyMC dependency to Aesara, etc.  Included in the rename is a rename of the `theanof` module to `aesaraf`.

~This most likely won't pass until the `conda-forge` rename (https://github.com/conda-forge/staged-recipes/pull/13825) and Arviz changes (https://github.com/arviz-devs/arviz/pull/1553) next release for Arviz is cut.~

### CODE REVIEW

**Comment by brandonwillard** [comment]
The Windows tests should pass once https://github.com/pymc-devs/aesara/pull/307 goes through.

File: README.rst [hunk 1]
```diff
@@ -15,13 +15,13 @@ Check out the `getting started guide <http://docs.pymc.io/notebooks/getting_star
 using Binder!
 For questions on PyMC3, head on over to our `PyMC Discourse <https://discourse.pymc.io/>`__ forum.
 
-The future of PyMC3 & Theano
+The future of PyMC3 & Aesara
```
Comment (michaelosthege) [inline README.rst]: The future is now - this whole chapter can probably be replaced with an "Updating" guide or so.

File: docs/source/Advanced_usage_of_Theano_in_PyMC3.rst [hunk 2]
```diff
@@ -4,20 +4,20 @@
...
 =================================
-Advanced usage of Theano in PyMC3
+Advanced usage of Aesara in PyMC3
 =================================
 
 Using shared variables
 ======================
 
-Shared variables allow us to use values in theano functions that are
+Shared variables allow us to use values in aesara functions that are
```
Comment (michaelosthege) [inline docs/source/Advanced_usage_of_Theano_in_PyMC3.rst]: What's the rule for capitalization in sentences like this one?

File: docs/source/PyMC3_and_Theano.rst [hunk 3]
```diff
@@ -4,24 +4,24 @@
...
 operations and linear algebra. When we define a PyMC3 model, we implicitly
-build up a Theano function from the space of our parameters to
+build up a Aesara function from the space of our parameters to
 their posterior probability density up to a constant factor. We then use
 symbolic manipulations of this function to also get access to its gradient.
 
-Note that the original developers have stopped maintaining Theano, so
-PyMC3 uses `Theano-PyMC <https://github.com/pymc-devs/Theano-PyMC>`_,
-a fork of Theano maintained by the PyMC3 developers.
+Note that the original developers have stopped maintaining Aesara, so
```
Comment (michaelosthege) [inline docs/source/PyMC3_and_Theano.rst]: Paragraph is outdated

File: docs/source/about.rst [hunk 4]
```diff
@@ -45,7 +45,7 @@ PyMC3 strives to make Bayesian modeling as simple and painless as possible,  all
...
+The third major version of PyMC has benefitted from being re-written from scratch. Substantial improvements in the user interface and performance have resulted from this. While PyMC2 relied on Fortran extensions (via f2py) for most of the computational heavy-lifting, PyMC3 leverages Aesara, a library from the Montréal Institute for Learning Algorithms (MILA), for array-based expression evaluation, to perform its computation. What this provides, above all else, is fast automatic differentiation, which is at the heart of the gradient-based sampling and optimization methods currently providing inference for probabilistic programming.
```
Comment (michaelosthege) [inline docs/source/about.rst]: Slightly outdated, specifically w.r.t. origin of Aesara

File: docs/source/about.rst [hunk 5]
```diff
@@ -90,7 +90,7 @@ plotting, csv table output, improved imputation syntax, and posterior
...
 
-In 2011, John Salvatier began thinking about implementing gradient-based MCMC samplers, and developed the ``mcex`` package to experiment with his ideas. The following year, John was invited by the team to re-engineer PyMC to accomodate Hamiltonian Monte Carlo sampling. This led to the adoption of Theano as the computational back end, and marked the beginning of PyMC3's development. The first alpha version of PyMC3 was released in June 2015. Over the following 2 years, the core development team grew to 12 members, and the first release, PyMC3 3.0, was launched in January 2017.
+In 2011, John Salvatier began thinking about implementing gradient-based MCMC samplers, and developed the ``mcex`` package to experiment with his ideas. The following year, John was invited by the team to re-engineer PyMC to accomodate Hamiltonian Monte Carlo sampling. This led to the adoption of Aesara as the computational back end, and marked the beginning of PyMC3's development. The first alpha version of PyMC3 was released in June 2015. Over the following 2 years, the core development team grew to 12 members, and the first release, PyMC3 3.0, was launched in January 2017.
```
Comment (michaelosthege) [inline docs/source/about.rst]: In the "History" it should probably remain as "Theano".

A new paragraph about PyMC taking over and renaming the backend would be great.

File: pymc3/model_graph.py [hunk 6]
```diff
@@ -13,19 +13,19 @@
...
+from aesara.compile import SharedVariable
+from aesara.graph.basic import walk
+from aesara.tensor.var import TensorVariable
 
 import pymc3 as pm
 
 from pymc3.model import ObservedRV
 from pymc3.util import get_default_varnames, get_var_name
 
+VarName = NewType("VarName", str)
```
Comment (michaelosthege) [inline pymc3/model_graph.py:27]: What's the background behind this change?

File: docs/source/Advanced_usage_of_Theano_in_PyMC3.rst (hunk 2 above)
Comment (brandonwillard) [reply docs/source/Advanced_usage_of_Theano_in_PyMC3.rst]: "Aesara/Theano" should be capitalized.

File: pymc3/model_graph.py (hunk 6 above)
Comment (brandonwillard) [reply pymc3/model_graph.py:27]: That's the correct way to create a distinct type from an existing type (see [here](https://mypy.readthedocs.io/en/stable/more_types.html#newtypes)).  Setting `VarName = str` does/offers nothing in terms of type checking.

**Comment by brandonwillard** [comment]
~[The only test failure](https://github.com/pymc-devs/pymc3/pull/4471/checks?check_run_id=1985394147#step:7:429) is suspiciously close to the expected relative error, and I can't reproduce it locally.  I also run Ubuntu and a very similar Conda environment, so I'm assuming this error is RNG-based.~

Setting the seed at the start of that one test resolves the error.

**Review by michaelosthege — APPROVED** [review]

**Pull request merged**